src/tls/TLSClient.fst

```diff
@@ -0,0 +1,298 @@
+  let kdf = PRF_TLS_1p2 kdf_label (HMAC CoreCrypto.SHA256) in
+  let gx = CommonDH.keygen (CommonDH.ECDH CoreCrypto.ECC_P256) in
+  let g = CommonDH.key_params gx in
+  let gy, gxy = CommonDH.dh_responder gx in
+  let msid = StandardMS (PMS.DHPMS(g, (CommonDH.share_of_key gx), (CommonDH.share_of_key gy), (PMS.ConcreteDHPMS gxy))) (er @| er) kdf in
+  ID12 TLS_1p2 msid kdf (AEAD CoreCrypto.AES_128_GCM CoreCrypto.SHA256) er er Client
+
+let encryptor_TLS12_AES_GCM_128_SHA256 key iv = 
+  let r = HyperHeap.root in
+  let w: writer id =
+    assume (~(authId id));
+    let seqn: HyperHeap.rref r seqn_t = ralloc r 0 in
+    let st: AEAD_GCM.state id Writer =
+      // The calls to [unsafe_coerce] are here because we're breaking
+      // abstraction, as both [key] and [iv] are declared as private types.
+      let key: AEAD_GCM.key id = key |> unsafe_coerce in
+      let iv: AEAD_GCM.iv id = iv |> unsafe_coerce in
+      let log: HyperHeap.rref r _ = ralloc r Seq.createEmpty in
+      let counter = ralloc r 0 in
+      AEAD_GCM.State key iv () counter
+    in
+    st
+  in
+  // StatefulLHAE.writer -> StatefulLHAE.state
+  w
+
+
+let decryptor_TLS12_AES_GCM_128_SHA256 key iv = 
+  let r = HyperHeap.root in
+  let r: reader id =
+    assume (~(authId id));
+    let seqn: HyperHeap.rref r seqn_t = ralloc r 0 in
+    let st: AEAD_GCM.state id Reader =
+      // The calls to [unsafe_coerce] are here because we're breaking
+      // abstraction, as both [key] and [iv] are declared as private types.
+      let key: AEAD_GCM.key id = key |> unsafe_coerce in
+      let iv: AEAD_GCM.iv id = iv |> unsafe_coerce in
+      let log: HyperHeap.rref r _ = ralloc r Seq.createEmpty in
+      let counter = ralloc r 0 in
+      AEAD_GCM.State key iv () counter
+    in
+    st
+  in
+  // StatefulLHAE.reader -> StatefulLHAE.state
+  r
+
+let encryptRecord_TLS12_AES_GCM_128_SHA256 w ct plain = 
+  let pv = TLS_1p2 in
+  let text = plain in
+  // StatefulPlain.adata id -> bytes
+  let ad: StatefulPlain.adata id = StatefulPlain.makeAD id ct in
+  // Range.frange -> Range.range
+  let rg: Range.frange id = 0, length text in
+  // DataStream.fragment -> DataStream.pre_fragment -> bytes
+  let f: DataStream.fragment id rg = text |> unsafe_coerce in
+  // LHAEPlain.plain -> StatefulPlain.plain -> Content.fragment
+  //NS: Not sure about the unsafe_coerce: but, it's presence clearly means that #id cannot be inferred
+  let f: LHAEPlain.plain id ad rg = Content.CT_Data #id rg f |> unsafe_coerce in
+  // StatefulLHAE.cipher -> StatefulPlain.cipher -> bytes
+  // FIXME: without the three additional #-arguments below, extraction crashes
+  StatefulLHAE.encrypt #id w ad rg f
+
+let decryptRecord_TLS12_AES_GCM_128_SHA256 rd ct cipher = 
+  let ad: StatefulPlain.adata id = StatefulPlain.makeAD id ct in
+  let (Some d) = StatefulLHAE.decrypt #id rd ad cipher in
+  Content.repr id d
+
+let hsbuf = alloc ([] <: list (hs_msg * bytes))
+
+type callbacks = FFICallbacks.callbacks
```

Comment:
out of curiosity, `callbacks` is an `int`... does it denote several callbacks or just one callback?